Issue 990 -- Moves compute source array to assembly

lsawade opened the pull request:
## 3D source to do list 

I'm copying this todo from PR to PR so that it can keep track what needs to happen still

- [x] Compute source array was now a part of assembly
- [x] Give the source `xi`, `gamma`, `medium tag` attributes 
- [x] Pull out the vector/tensor return out of the compute_source_array
- [x] Move `{include,src}/source` to `core/specfem/source`
- [x] Split sources into vector and tensor ABC's 
  - declare get_vector and get_tensor member functions, which are overriden by the derived class.
  - this also implements a support-based `set_medium_tag` function which fails if the source is placed in the wrong medium.
- [x] Implement the tests for compute_source_array functions
  + Add test that expects failure if the wrong medium is assigned (add to pre-existing tests)  
  + `compute_source_array(..., Vector)` (this test is simple)
  + `compute_source_array(..., Tensor)` (this test is not simple)
- [x] Dimension template the source for 2D 
- [x] Implement 3D source
- [x] Implement 3D source readers 
- [x] Consolidating source types 
- [x] Waiting for 3D locate_point (Started [on branch issue-1057]
- [x] Interpolate function  
- [ ] Implement 3D assembly sources (Underway, but requires interpolate function)

### Seperate to merge this for now:

- [ ] Update documentation

## Notes

Tags in comments for future updates:

- EXTENT_IMPL_CHANGE -> after upgrade to kokkos 4.7
- CPP20_CHANGE -> Syntax needs update for C++20 

## Issue Number

Updates #990

## Checklist

Please make sure to check developer documentation on specfem docs.

- [x] I ran the code through pre-commit to check style
- [x] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [x] I have added labels to the PR (see right hand side of the PR page)
- [x] My code passes all the integration tests
- [x] I have added sufficient unittests to test my changes
- [x] I have added/updated documentation for the changes I am proposing
- [x] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

icui commented on src/source/force_source.cpp:
The `almost_equal` is no longer used right? It has been switched to `utilities::almost_equal`

Rohit-Kakodkar commented on core/specfem/assembly/compute_source_array/dim2/impl/external_source_array.cpp:
```diff
@@ -0,0 +1,115 @@
+        source_array(2, iz, ix) = hlagrange;
+      } else {
+        std::ostringstream message;
+        message << "Source array computation not implemented for element type: "
+                << specfem::element::to_string(el_type) << " [" << __FILE__
+                << ":" << __LINE__ << "]";
+        auto message_str = message.str();
+        Kokkos::abort(message_str.c_str());
```
Abort with message?

Rohit-Kakodkar reviewed (changes requested):
Minor changes

lsawade commented on core/specfem/assembly/compute_source_array.hpp:
```diff
@@ -0,0 +1,57 @@
+  bool computed_source = false;
+
+  // Loop through the function that compute the source.
+  for (const auto &compute_source_array :
```
you mean using dextents for HostView?

lsawade commented:
Since your last review,  
- I added local coordinates `xi`, and `gamma`, `element index`, `medium_tag`. This allows us to
- remove redundant locating of sources in `source_medium` (for each medium), and the again in `compute_source_array`
- I added solutions to the sources in the mesh IO tests, so that the location and medium_tag setting is being tested.
